### PR TITLE
WIP: Add volume info and volume list for CLI

### PR DESCRIPTION
**Pull request by devyanikota**
Still working on it.
there are json decoding issues.

> json: cannot unmarshal number into Go value of type string

PR is WIP.

### CODE REVIEW

**Comment by devyanikota** [comment]
Resolved unmarshalling issues. @aravindavk can you please check once. Thanks.

File: cli/cmd/volume.go [hunk 1]
```diff
@@ -152,16 +154,44 @@ var volumeResetCmd = &cobra.Command{
...
 
+func volumeInfoHandler2(cmd *cobra.Command)(api.Volinfo, error) {
+	var vols api.Volinfo
+	validateNArgs(cmd, 0, 1)
+	volname := ""
+	if len(cmd.Flags().Args()) > 0 {
+		volname = cmd.Flags().Args()[0]
+	}
+	if volname == ""{
+		vols_, err := client.VolumeAll()
```
Comment (aravindavk) [inline cli/cmd/volume.go]: same function can be used here

client.Volumes("")

File: cli/cmd/volume.go [hunk 2]
```diff
@@ -152,16 +154,44 @@ var volumeResetCmd = &cobra.Command{
 	},
 }
 
+func volumeInfoHandler2(cmd *cobra.Command)(api.Volinfo, error) {
```
Comment (aravindavk) [inline cli/cmd/volume.go]: add one more argument to decide info or list. Example "isInfo bool"

File: cli/cmd/volume.go [hunk 3]
```diff
@@ -152,16 +154,44 @@ var volumeResetCmd = &cobra.Command{
...
+		vols_, err := client.VolumeAll()
+		if err != nil {
+			log.Println("No volumes present")
+			failure(fmt.Sprintf("Error getting Volumes list %s", err.Error()), 1)
+		}
+		fmt.Println(vols_)
+	}else{
+		vols, _ = client.Volumes(volname)
+	}
+	return vols, nil
```
Comment (aravindavk) [inline cli/cmd/volume.go]: Add display functionality here.

if isInfo{
// Add Info display for loop
} else {
// Add list display as table with heading ID  Name
}

File: cli/cmd/volume.go [hunk 4]
```diff
@@ -152,16 +154,44 @@ var volumeResetCmd = &cobra.Command{
...
 	Short: helpVolumeInfoCmd,
 	Run: func(cmd *cobra.Command, args []string) {
-		validateNArgs(cmd, 0, 1)
-		volname := "all"
-		if len(cmd.Flags().Args()) > 0 {
-			volname = cmd.Flags().Args()[0]
-		}
-		fmt.Println("INFO:", volname)
+		vols, err := volumeInfoHandler2(cmd)
+		fmt.Println(vols)
```
Comment (aravindavk) [inline cli/cmd/volume.go]: Don't return vols, handle err

err := volumeInfoHandler2(cmd, true)

File: cli/cmd/volume.go [hunk 5]
```diff
@@ -152,16 +154,44 @@ var volumeResetCmd = &cobra.Command{
...
+		// 		fmt.Println("Volume Name: ", vol.Name)
+		// }
+	},
+}
+
+var volumeListCmd = &cobra.Command{
+	Use:   "list",
+	Short: helpVolumeListCmd,
+	Run: func(cmd *cobra.Command, args []string) {
+		volumeInfoHandler2(cmd)
```
Comment (aravindavk) [inline cli/cmd/volume.go]: same as above

err := volumeInfoHandler2(cmd, false)

File: pkg/restclient/volume.go [hunk 6]
```diff
@@ -15,12 +15,19 @@ func (c *Client) VolumeCreate(req api.VolCreateReq) (api.Volinfo, error) {
 }
 
 // Volumes returns list of all volumes
-func (c *Client) Volumes() (api.VolList, error) {
-	var vols api.VolList
+func (c *Client) VolumeAll() ([]api.Volinfo, error) {
```
Comment (aravindavk) [inline pkg/restclient/volume.go]: New func not required, modify the Volumes

File: pkg/restclient/volume.go [hunk 7]
```diff
@@ -15,12 +15,19 @@ func (c *Client) VolumeCreate(req api.VolCreateReq) (api.Volinfo, error) {
...
+	var vols []api.Volinfo
 	err := c.get("/v1/volumes", nil, http.StatusOK, &vols)
 	return vols, err
 }
 
+func (c *Client) Volumes(volname string) (api.Volinfo, error) {
```
Comment (aravindavk) [inline pkg/restclient/volume.go]: Return list of Vols

func (c *Client) Volumes(volname string) ([]api.Volinfo, error)

File: pkg/restclient/volume.go [hunk 8]
```diff
@@ -15,12 +15,19 @@ func (c *Client) VolumeCreate(req api.VolCreateReq) (api.Volinfo, error) {
...
+	var vols api.Volinfo
```
Comment (aravindavk) [inline pkg/restclient/volume.go]: use []api.Volinfo

File: pkg/restclient/volume.go [hunk 9]
```diff
@@ -15,12 +15,19 @@ func (c *Client) VolumeCreate(req api.VolCreateReq) (api.Volinfo, error) {
...
+	url := fmt.Sprintf("/v1/volumes/%s", volname)
```
Comment (aravindavk) [inline pkg/restclient/volume.go]: if volname == ""{
url := fmt.Sprintf("/v1/volumes")
err := c.get(url, nil, http.StatusOK, &vols)
return vols, err
} else {
var vol api.Volinfo
url := fmt.Sprintf("/v1/volumes/%s", volname)
err := c.get(url, nil, http.StatusOK, &vol)
return []api.Volinfo{vol}, err
}

File: cli/cmd/volume.go (hunk 2 above)
Comment (devyanikota) [reply cli/cmd/volume.go]: good idea :) implemented the same.

File: cli/cmd/volume.go (hunk 1 above)
Comment (devyanikota) [reply cli/cmd/volume.go]: Rectified!

File: cli/cmd/volume.go (hunk 3 above)
Comment (devyanikota) [reply cli/cmd/volume.go]: Done!

File: pkg/restclient/volume.go (hunk 7 above)
Comment (devyanikota) [reply pkg/restclient/volume.go]: nod nod. rectified :)

File: pkg/restclient/volume.go (hunk 9 above)
Comment (devyanikota) [reply pkg/restclient/volume.go]: Made the suggested changes! Please review the same. Thanks.

**Comment by devyanikota** [comment]
@aravindavk edited the tests as well. Can you please review this patch. Thanks.

**Comment by devyanikota** [comment]
Also can someone check why its not building. Jenkins job is stuck since last night. Thanks.

**Pull request closed without merging**